src/compiler/checker.ts

```diff
@@ -22396,11 +22397,38 @@ namespace ts {
         }
 
         function getAccessedPropertyName(access: AccessExpression | BindingElement): __String | undefined {
-            let propertyName;
-            return access.kind === SyntaxKind.PropertyAccessExpression ? access.name.escapedText :
-                access.kind === SyntaxKind.ElementAccessExpression && isStringOrNumericLiteralLike(access.argumentExpression) ? escapeLeadingUnderscores(access.argumentExpression.text) :
-                access.kind === SyntaxKind.BindingElement && (propertyName = getDestructuringPropertyName(access)) ? escapeLeadingUnderscores(propertyName) :
-                undefined;
+            if (isPropertyAccessExpression(access)) {
+                return access.name.escapedText;
+            }
+            if (isElementAccessExpression(access)) {
+                return tryGetElementAccessExpressionName(access);
+            }
+            if (isBindingElement(access)) {
+                const name = getDestructuringPropertyName(access);
+                return name ? escapeLeadingUnderscores(name) : undefined;
+            }
+            return undefined;
+        }
+
+        function tryGetElementAccessExpressionName(node: ElementAccessExpression) {
+            if (isStringOrNumericLiteralLike(node.argumentExpression)) {
+                return escapeLeadingUnderscores(node.argumentExpression.text);
+            }
+            if (isIdentifier(node.argumentExpression)) {
+                const symbol = getResolvedSymbol(node.argumentExpression);
+                if (!isConstVariable(symbol)) return undefined;
+
+                const declaration = symbol.valueDeclaration;
+                if (!(declaration && hasOnlyExpressionInitializer(declaration))) return undefined;

```

Comment:
Ooo, we may also want to handle using the type annotation on the declaration if it has one, rather than the initializer's type.